packages/router/src/adapters/contract/contract.ts

```diff
@@ -130,53 +89,100 @@ export const cancelAndFullfillSanitationCheck = async (
-  }
-
-  // transaction should be prepared before fulfill
-  if (variantTransactionDigest === HashZero) {
-    const error = new Error("Transaction isn't prepared yet");
-    logger.error(
-      "FAILED cancelAndFullfillSanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT: Transaction isn't prepared yet",
-      requestContext,
-      methodContext,
-      jsonifyError(error),
-      { transactionData, chainId },
-    );
-    throw error;
+  if (functionCall === "prepare") {
+    // variantTransactionDigest exist then transaction is already prepared
+    if (variantTransactionDigest !== HashZero) {
+      const error = new Error("Transaction is already prepared");
+      logger.error(
+        `FAILED ${functionCall} sanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT`,
+        requestContext,
+        methodContext,
+        jsonifyError(error),
+        { transactionData, chainId },
+      );
+      throw error;
+    }
+  } else {
+    const expectedVariantDigest = getVariantTransactionDigest({
+      amount: transactionData.amount,
+      expiry: transactionData.expiry,
+      preparedBlockNumber: transactionData.preparedBlockNumber,
+    });
+
+    if (expectedVariantDigest === variantTransactionDigest) {
+      // All is good, no issues
+      return;
+    }
+
+    // transaction should be prepared before fulfill
+    if (variantTransactionDigest === HashZero) {
+      const error = new Error("Transaction isn't prepared yet");
+      logger.error(
+        `FAILED ${functionCall} sanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT`,
+        requestContext,
+        methodContext,
+        jsonifyError(error),
+        { transactionData, chainId },
+      );
+      throw error;
+    }
+
+    // transaction is already fulfilled
+    // get expected fulfilled/cancelled variant hash
+    const fulfilledOrCancelledVariant = getVariantTransactionDigest({
+      amount: transactionData.amount,
+      expiry: transactionData.expiry,
+      preparedBlockNumber: 0,
+    });
+
+    if (variantTransactionDigest === fulfilledOrCancelledVariant) {
+      const error = new Error("Transaction is already fulfilled or canceled");
+      logger.error(
+        `FAILED ${functionCall} sanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT`,
+        requestContext,
+        methodContext,
+        jsonifyError(error),
+        { transactionData, chainId },
+      );
+      throw error;
+    }
+
+    if (functionCall === "cancel" && chainId === transactionData.sendingChainId) {
+      const receivingChainNxtpContractAddress = getContractAddress(transactionData.receivingChainId);
+
+      const receivingChainVariantTransactionDigest = await txService.readTx({
+        chainId: transactionData.receivingChainId,
+        to: receivingChainNxtpContractAddress,
+        data: encodeVariantTransactionData,
+      });
+
+      if (receivingChainVariantTransactionDigest === HashZero) {
+        // cancel is allowed when no transaction is prepared
```

Comment:
`// cancel is allowed when no transaction is prepared on receiving chain`
sorry for these nits just trying to be super specific here